File: sdks/javascript/entry/cache.entry.ts
export type CacheSource = "API" | "CACHE";

export interface CacheMetadata {
  source: CacheSource;
  created_at: number;
  expires_at: number;
  ttl_ms: number;
  schema_version: string;
  data_version: string;
  spec_checksum: string;
  cache_namespace: string;
  compressed: boolean;
}

export interface CacheEntry<T = unknown> {
  cache_key: string;
  data: T;
  metadata: CacheMetadata;
}

export interface CacheMetadataParityExpectation {
  schema_version: string;
  spec_checksum: string;
  cache_namespace: string;
}

/**
 * Deterministic parity validator for metadata fields required by LCP.
 * Returns false on any mismatch or malformed value.
 */
export function isCacheMetadataParityValid(
  metadata: CacheMetadata,
  expected: CacheMetadataParityExpectation
): boolean {
  if (!metadata) return false;

  if (metadata.schema_version !== expected.schema_version) return false;
  if (metadata.spec_checksum !== expected.spec_checksum) return false;
  if (metadata.cache_namespace !== expected.cache_namespace) return false;

  if (metadata.source !== "API" && metadata.source !== "CACHE") return false;
  if (!Number.isFinite(metadata.created_at) || metadata.created_at < 0) return false;
  if (!Number.isFinite(metadata.expires_at) || metadata.expires_at < 0) return false;
  if (!Number.isFinite(metadata.ttl_ms) || metadata.ttl_ms < 0) return false;
  if (metadata.expires_at !== metadata.created_at + metadata.ttl_ms) return false;
  if (!metadata.data_version) return false;

  return true;
}

/**
 * Build immutable metadata from API-authoritative TTL.
 */
export function createCacheMetadata(input: {
  source: CacheSource;
  createdAt: number;
  ttlMs: number;
  schemaVersion: string;
  dataVersion: string;
  specChecksum: string;
  cacheNamespace: string;
  compressed?: boolean;
}): CacheMetadata {
  const expiresAt = input.createdAt + input.ttlMs;
  return Object.freeze({
    source: input.source,
    created_at: input.createdAt,
    expires_at: expiresAt,
    ttl_ms: input.ttlMs,
    schema_version: input.schemaVersion,
    data_version: input.dataVersion,
    spec_checksum: input.specChecksum,
    cache_namespace: input.cacheNamespace,
    compressed: Boolean(input.compressed)
  });
}

File: sdks/flutter/lib/src/entry/cache_entry.dart
enum CacheSource { api, cache }

class CacheMetadata {
  CacheMetadata({
    required this.source,
    required this.createdAt,
    required this.expiresAt,
    required this.ttlMs,
    required this.schemaVersion,
    required this.dataVersion,
    required this.specChecksum,
    required this.cacheNamespace,
    required this.compressed,
  });

  final CacheSource source;
  final int createdAt;
  final int expiresAt;
  final int ttlMs;
  final String schemaVersion;
  final String dataVersion;
  final String specChecksum;
  final String cacheNamespace;
  final bool compressed;

  Map<String, Object> toJson() => {
        'source': source == CacheSource.api ? 'API' : 'CACHE',
        'created_at': createdAt,
        'expires_at': expiresAt,
        'ttl_ms': ttlMs,
        'schema_version': schemaVersion,
        'data_version': dataVersion,
        'spec_checksum': specChecksum,
        'cache_namespace': cacheNamespace,
        'compressed': compressed,
      };
}

class CacheEntry<T> {
  CacheEntry({
    required this.cacheKey,
    required this.data,
    required this.metadata,
  });

  final String cacheKey;
  final T data;
  final CacheMetadata metadata;
}

class CacheMetadataParityExpectation {
  const CacheMetadataParityExpectation({
    required this.schemaVersion,
    required this.specChecksum,
    required this.cacheNamespace,
  });

  final String schemaVersion;
  final String specChecksum;
  final String cacheNamespace;
}

bool isCacheMetadataParityValid(
  CacheMetadata metadata,
  CacheMetadataParityExpectation expected,
) {
  if (metadata.schemaVersion != expected.schemaVersion) return false;
  if (metadata.specChecksum != expected.specChecksum) return false;
  if (metadata.cacheNamespace != expected.cacheNamespace) return false;
  if (metadata.createdAt < 0 || metadata.expiresAt < 0 || metadata.ttlMs < 0) {
    return false;
  }
  if (metadata.expiresAt != metadata.createdAt + metadata.ttlMs) return false;
  if (metadata.dataVersion.isEmpty) return false;
  return true;
}

CacheMetadata createCacheMetadata({
  required CacheSource source,
  required int createdAt,
  required int ttlMs,
  required String schemaVersion,
  required String dataVersion,
  required String specChecksum,
  required String cacheNamespace,
  bool compressed = false,
}) {
  final expiresAt = createdAt + ttlMs;
  return CacheMetadata(
    source: source,
    createdAt: createdAt,
    expiresAt: expiresAt,
    ttlMs: ttlMs,
    schemaVersion: schemaVersion,
    dataVersion: dataVersion,
    specChecksum: specChecksum,
    cacheNamespace: cacheNamespace,
    compressed: compressed,
  );
}
